### ISSUE TITLE
Shared/global context, scalars, schema directives

### ISSUE THREAD

**Issue by robbertkl** [Urigo/graphql-modules#720]
Hi, I'm still fairly new to graphql-modules and there's still 1 thing bugging me: before graphql-modules, I was used to setting my ApolloServer `context` with some shared objects (request auth info, db connection pool, etc), but with modules context seems to work a bit different.

First of all, the context I set in ApolloServer does not end up as context in the modules' resolvers. Instead, the only context that is available is from the module's own context function and of its imports. I did notice the ApolloServer context is available as `session` (the first argument) in the module's context function.

I can "fix" this by doing `context: context => context,` in each of my modules, but this becomes tedious. Another solution is to add the context in 1 or more separate modules (either creating the context there or doing the `context => context` trick) and include it/them in my other modules. This, again, becomes tedious.

I've tried adding the context to my "root" module (`AppModule`), but since this imports the other modules, its context gets added after the other modules, so it doesn't work either.

Aside from context, I'd also like to have some shared custom scalars and directives to be available to all modules. This suffers from the same problem (has to be imported manually every time).

The only ok-ish solution is to have some kind of CommonModule which combines (imports) all the shared components (context, typeDefs, resolvers) and import it it all of my modules.

Finally there's the thing with schema directives, which I add using `SchemaDirectiveVisitor.visitSchemaDirectives()`. Any field resolvers I add in a `SchemaDirectiveVisitor`, do get the ApolloServer context, and not the context from the module where the directive was used. This means I have to make sure the same context is available through the ApolloServer config.

Here's part of the code that sets up the root module and ApolloServer:

```js
const AppModule = new GraphQLModule({
  imports: Object.values(modules),
});

const { schema, schemaDirectives } = AppModule;
SchemaDirectiveVisitor.visitSchemaDirectives(schema, schemaDirectives);

const apollo = new ApolloServer({
  schema,
```

Basically, what I'd like to accomplish is a place to define schema (typedefs with custom scalars & directives, resolvers) and context that are automatically shared in all modules. Is this somehow possible? Am I thinking in the wrong way? Should I look into DI? I'd rather not go into Providers / `injector` as my project is still fairly simple. Hope someone can help, thanks!

**Comment by johnrix** [comment]
Experiencing precisely the same issues today also.

**Comment by johnrix** [comment]
@robbertkl, a workaround for the issue of shared context data that I found is to include a `global` (or otherwise named) object in the result of the context function provided when creating your ApolloServer instance. This will be visible (and can be manipulated) on the `session` object supplied to the context function for each GraphQLModule instance, and also within dataSources or wherever else context appears:

```
const server = new ApolloServer({ 
    schema,
    context: session => {
        return {
            ...session,
            global: {
                host: session.req.headers.host,
                env: process.env.NODE_ENV,
                ...
            }
        }
    },
    dataSources: () => {
        return {
            ...
        }
    }, 
});
```

Not an ideal solution, but gets around this particular issue until a better option surfaces!

**Comment by robbertkl** [comment]
@johnrix Yeah, that's sort of what I'm doing already. I have a few workarounds mentioned in the issue (like doing `context: context => context` or import a separate context module), but it's not ideal.

**Comment by johnrix** [comment]
I did indeed use your `context: context => context` trick also!

**Comment by IsenrichO** [comment]
I think this is something that would help improve the "quality of life" for many users of this already great library? Support for shared/global contexts would be excellent. Is this something being worked on and/or would contributions for the feature be welcome?

**Comment by c0ldra1n** [comment]
This is a very intuitive and needed feature. Injectors are fine, but this would make data transfer/transform much easier.

**Comment by fuchsm** [comment]
Hi, 

I was also running into the very same kind of problem, took me a while to figure out that context doesn't work the usual way here.

@johnrix,

I was also trying to do this:

```
const server = new ApolloServer({ 
    schema,
    context: session => {
        return {
            ...session,
            global: {
                host: session.req.headers.host,
                env: process.env.NODE_ENV,
                ...
            }
        }
    },
});
```

But then how to access the global vars within the module? You said:  

> This will be visible (and can be manipulated) on the session object supplied to the context function for each GraphQLModule instance

Unfortunately, I haven't found a way to access it from within the module, at least global.host does not show up as additional session data.

I had also figured out to put `context: (session) => session` in each module, at least If I don't have it there, I cannot even access session. 

At the moment, all I want is one (or more) additional global context (besides session) accessible within all of my modules.

**Comment by fridaystreet** [comment]
Hi guys,

Thought I might share my findings hopefully may be of some help. In regards to the directives, It took me a while to figure this out too as I kept hitting a brick wall trying to use Apollo's schema directive implementation. Then I found this article on example to implementing authorisation in graphql-modules, which shone the light on how to implement schema directives through resolver composition in graphql-modules. While it starts off showing how to use composition the way shown in the docs, it goes on at the end to show how it can be implemented with schema directives.

https://medium.com/the-guild/authentication-and-authorization-in-graphql-and-how-graphql-modules-can-help-fadc1ee5b0c2

I'd already seen a webinar (sorry can't remember link) on using schema directives for authorization, but the graphql-modules docs don't have any reference to schema directives regarding resolver composition so couldn't join the dots. It's a simplistic example which only does field level, but it's easy to extend to other levels/types of schema. And it allows your directive resolvers to be native providers in graphql-modules so they get all the fruit and context your normal resolvers/providers get.

We've implemented it using a hierarchy import type pattern. So each module can be decoupled and responsible for it's own directives. for example

ServerModule -imports-> API -imports-> Comments -imports-> User -imports-> Auth -imports-> [Services, DataSources]  

We started out with each module implementing its own resolver composition into it's graphql-modules resolver-composition property, but then realised the composition is only run against the schema within that module. (eg if auth as an @auth directive, but we want to use that on 'postComment' mutation in comment module, it wouldn't be applied. So we opted to export the directive resolvers as a separate export as well from each of the modules index files and collate them up the tree. So the final outcome being API exports all of the directive resolvers  which our main server module imports and passes to the top level graphql-module through it's resolverComposition, which means it iterates the entire schema and attaches directive resolvers from all modules, so they can be used anywhere.

It sounds like a bit of a workaround and a tradeoff, but it's working nicely for us. Would be nice to see this sort of rollup of resolver composition baked into graphql modules down the track.

**Comment by mikew** [comment]
The docs state that the first argument pass to a graphql-modules context function is the session, but it's not. It's much less clear than that:

it's what your apollo server's context returns. if you're returning an object with a key named `session`, it's that, and anything else in the object is discarded

**Comment by kamilkisiela** [comment]
We released a new major version (v1.0) of GraphQL Modules. Here's the [migration guide](https://graphql-modules.com/docs/next/recipes/migration).

The new version solves the issue by not enforcing any context wrapping logic, so I'm closing the thread.

Please try to upgrade. If you can't and still want to use the legacy version, please open a new issue.
